File: github-tools/deployments_cleaner.go
package github_tools

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"
)

//[doc](https://docs.github.com/en/rest/deployments/deployments?apiVersion=2022-11-28)

// Replace with your GitHub username
var username = ""

// Replace with your GitHub token,[personal access token](https://github.com/settings/tokens)
var token = ""

// Replace with your GitHub repositoryName
var repositoryName = ""

var baseURL = "https://api.github.com/repos/" + username + "/" + repositoryName
var deploymentsURL = baseURL + "/deployments"
var statusesURL = baseURL + "/deployments/%d/statuses"

// Function to get deployments
func getDeployments() (*http.Response, error) {
	req, err := http.NewRequest("GET", deploymentsURL, nil)
	if err != nil {
		return nil, err
	}
	req.SetBasicAuth(username, token)

	client := &http.Client{}
	return client.Do(req)
}

// Function to mark a deployment as inactive
func inactiveDeployment(deploymentId int) (*http.Response, error) {
	url := fmt.Sprintf(statusesURL, deploymentId)
	data := map[string]string{"state": "inactive"}
	jsonData, _ := json.Marshal(data)

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, err
	}
	req.SetBasicAuth(username, token)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/vnd.github.ant-man-preview+json")

	client := &http.Client{}
	return client.Do(req)
}

// Function to delete a deployment
func deleteDeployment(deploymentId int) (*http.Response, error) {
	inactivateResponse, err := inactiveDeployment(deploymentId)
	if err != nil {
		return nil, err
	}
	defer func(Body io.ReadCloser) {
		_ = Body.Close()
	}(inactivateResponse.Body)

	if inactivateResponse.StatusCode == 201 {
		fmt.Printf("Deployment %d marked as inactive.\n", deploymentId)

		url := fmt.Sprintf("%s/%d", deploymentsURL, deploymentId)
		req, err := http.NewRequest("DELETE", url, nil)
		if err != nil {
			return nil, err
		}
		req.SetBasicAuth(username, token)

		client := &http.Client{}
		return client.Do(req)
	} else {
		fmt.Printf("Failed to update deployment %d. Status Code: %d\n", deploymentId, inactivateResponse.StatusCode)
		return nil, fmt.Errorf("failed to update deployment")
	}
}

// Function to delete all deployments found in this query
func deleteDeploymentsThisQuery() bool {
	response, err := getDeployments()
	if err != nil {
		log.Fatal(err)
	}
	defer func(Body io.ReadCloser) {
		_ = Body.Close()
	}(response.Body)

	if response.StatusCode != 200 {
		fmt.Printf("Failed to retrieve deployments. Status Code: %d\n", response.StatusCode)
		return false
	}

	body, err := io.ReadAll(response.Body)
	if err != nil {
		log.Fatal(err)
	}

	var deployments []map[string]interface{}
	if err := json.Unmarshal(body, &deployments); err != nil {
		log.Fatal(err)
	}

	if len(deployments) == 0 {
		return false
	}

	for _, deployment := range deployments {
		id, ok := deployment["id"].(float64)
		if !ok {
			continue
		}
		_, err := deleteDeployment(int(id))
		if err != nil {
			log.Fatal(err)
		}
	}

	return true
}

// DeleteAllDeployments Function to delete all deployments
func DeleteAllDeployments() {
	for {
		if !deleteDeploymentsThisQuery() {
			break
		}
		time.Sleep(time.Second * 2)
	}
}

File: github-tools/deployments_cleaner.py
import requests
import time

# Replace with your GitHub username
username = ''

# Replace with your GitHub token, [personal access token](https://github.com/settings/tokens)
token = ''

# Replace with your GitHub repository name
repository_name = ''

base_url = f'https://api.github.com/repos/{username}/{repository_name}'
deployments_url = f'{base_url}/deployments'
statuses_url = f'{base_url}/deployments/{{}}/statuses'


# Function to get deployments
def get_deployments():
    response = requests.get(deployments_url, auth=(username, token))
    if response.status_code == 200:
        return response.json()
    else:
        response.raise_for_status()


# Function to mark a deployment as inactive
def inactive_deployment(deployment_id):
    url = statuses_url.format(deployment_id)
    data = {"state": "inactive"}
    response = requests.post(url, json=data, auth=(username, token),
                             headers={"Content-Type": "application/json",
                                      "Accept": "application/vnd.github.ant-man-preview+json"})
    if response.status_code == 201:
        print(f"Deployment {deployment_id} marked as inactive.")
        return True
    else:
        print(f"Failed to update deployment {deployment_id}. Status Code: {response.status_code}")
        return False


# Function to delete a deployment
def delete_deployment(deployment_id):
    if inactive_deployment(deployment_id):
        url = f'{deployments_url}/{deployment_id}'
        response = requests.delete(url, auth=(username, token))
        return response
    else:
        return None


# Function to delete all deployments found in this query
def delete_deployments_this_query():
    try:
        deployments = get_deployments()
    except requests.HTTPError as e:
        print(f"Failed to retrieve deployments. Error: {e}")
        return False

    if not deployments:
        return False

    for deployment in deployments:
        deployment_id = deployment.get("id")
        if deployment_id is not None:
            response = delete_deployment(int(deployment_id))
            if response and response.status_code != 204:
                print(f"Failed to delete deployment {deployment_id}. Status Code: {response.status_code}")

    return True


# Function to delete all deployments
def delete_all_deployments():
    while True:
        try:
            if not delete_deployments_this_query():
                break
            time.sleep(2)
        except Exception as e:
            print(f"An error occurred: {e}")
            break
